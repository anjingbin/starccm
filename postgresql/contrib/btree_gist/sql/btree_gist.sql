--
-- first, define the datatype.  Turn off echoing so that expected file
-- does not depend on contents of seg.sql.
--
\set ECHO none
\i btree_gist.sql
\set ECHO all

create table inttmp (b int4);

\copy inttmp from 'data/test_btree.data'

create table tstmp ( t timestamp without time zone );

\copy tstmp from 'data/test_btree_ts.data'

-- without idx

select count(*) from inttmp where b <=10;

select count(*) from tstmp where t < '2001-05-29 08:33:09';

-- create idx

create index aaaidx on inttmp using gist ( b );

create index tsidx on tstmp using gist ( t );

--with idx

set enable_seqscan=off;

select count(*) from inttmp where b <=10;

select count(*) from tstmp where t < '2001-05-29 08:33:09';

