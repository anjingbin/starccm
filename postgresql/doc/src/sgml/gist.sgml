<Chapter Id="gist">
<DocInfo>
<AuthorGroup>
<Author>
<FirstName>Gene</FirstName>
<Surname>Selkov</Surname>
</Author>
</AuthorGroup>
<Date>Transcribed 1998-02-19</Date>
</DocInfo>
<Title>GiST Indexes</Title>

<Para>
The information about GIST is at
 <ULink url="http://GiST.CS.Berkeley.EDU:8000/gist/">http://GiST.CS.Berkeley.EDU:8000/gist/</ULink>

with more on different indexing and sorting schemes at
<ULink url="http://s2k-ftp.CS.Berkeley.EDU:8000/personal/jmh/">http://s2k-ftp.CS.Berkeley.EDU:8000/personal/jmh/</ULink>.

And there is more interesting reading at
<ULink url="http://epoch.cs.berkeley.edu:8000/">http://epoch.cs.berkeley.edu:8000/</ULink> and
<ULink url="http://www.sai.msu.su/~megera/postgres/gist/">http://www.sai.msu.su/~megera/postgres/gist/</ULink>.
</para>

<Para>
<Note>
<Title>Author</Title>
<Para>
This extraction from an email sent by 
Eugene Selkov, Jr. (<email>selkovjr@mcs.anl.gov</email>)
contains good information
on GiST. Hopefully we will learn more in the future and update this information.
- thomas 1998-03-01
</Para>
</Note>
</para>
<Para>
Well, I can't say I quite understand what's going on, but at least
I (almost) succeeded in porting GiST examples to linux. The GiST access
method is already in the postgres tree (<FileName>src/backend/access/gist</FileName>).
</para>
<Para>
<ULink url="ftp://s2k-ftp.cs.berkeley.edu/pub/gist/pggist/pggist.tgz">Examples at Berkeley</ULink>
come with an overview of the methods and demonstrate spatial index 
mechanisms for 2D boxes, polygons, integer intervals and text
(see also <ULink url="http://gist.cs.berkeley.edu:8000/gist/">GiST at Berkeley</ULink>).
In the box example, we
are supposed to see a performance gain when using the GiST index; it did
work for me but I do not have a reasonably large collection of boxes
to check that. Other examples also worked, except polygons: I got an 
error doing

<ProgramListing>
test=> CREATE INDEX pix ON polytmp
test-> USING GIST (p:box gist_poly_ops) WITH (ISLOSSY);
ERROR:  cannot open pix

(PostgreSQL 6.3               Sun Feb  1 14:57:30 EST 1998)
</ProgramListing>
</para>
<Para>
I could not get sense of this error message; it appears to be something
we'd rather ask the developers about (see also Note 4 below). What I
would suggest here is that someone of you linux guys (linux==gcc?) fetch the 
original sources quoted above and apply my patch (see attachment) and 
tell us what you feel about it. Looks cool to me, but I would not like 
to hold it up while there are so many competent people around.
</para>
<Para>
A few notes on the sources:
</para>
<Para>
1. I failed to make use of the original (HP-UX) Makefile and rearranged
   the Makefile from the ancient postgres95 tutorial to do the job. I tried
   to keep it generic, but I am a very poor makefile writer -- just did
   some monkey work. Sorry about that, but I guess it is now a little
   more portable that the original makefile.
</para>
<Para>
2. I built the example sources right under pgsql/src (just extracted the
   tar file there). The aforementioned Makefile assumes it is one level
   below pgsql/src (in our case, in pgsql/src/pggist).
</para>
<Para>
3. The changes I made to the *.c files were all about #include's,
   function prototypes and typecasting. Other than that, I just threw 
   away a bunch of unused vars and added a couple parentheses to please
   gcc. I hope I did not screw up too much :)
</para>
<Para>
4. There is a comment in polyproc.sql:

<ProgramListing>
-- -- there's a memory leak in rtree poly_ops!!
-- -- CREATE INDEX pix2 ON polytmp USING RTREE (p poly_ops);
</ProgramListing>

   Roger that!! I thought it could be related to a number of
   <ProductName>PostgreSQL</ProductName> versions
   back and tried the query. My system went nuts and I had to shoot down
   the postmaster in about ten minutes.
</para>

<Para>
I will continue to look into GiST for a while, but I would also
appreciate
more examples of R-tree usage.
</para>
</Chapter>
