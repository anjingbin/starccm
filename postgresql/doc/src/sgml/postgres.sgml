<!--
$Header: /home/hjcvs/OB-CCM-1.0/postgresql/doc/src/sgml/postgres.sgml,v 1.2 2004/07/05 08:55:12 SuLiang Exp $
-->

<!doctype set PUBLIC "-//OASIS//DTD DocBook V3.1//EN" [

<!entity % version SYSTEM "version.sgml">
%version;
<!entity % filelist SYSTEM "filelist.sgml">
%filelist;

<!entity % set-of-books "INCLUDE">
<!entity % single-book "IGNORE">

<!entity admin      SYSTEM "admin.sgml">
<!entity developer  SYSTEM "developer.sgml">
<!entity programmer SYSTEM "programmer.sgml">
<!entity reference  SYSTEM "reference.sgml">
<!entity tutorial   SYSTEM "tutorial.sgml">
<!entity user       SYSTEM "user.sgml">

]>

<set id="postgres">
 <title>PostgreSQL &version; Documentation</title>

 <setinfo>
  <corpauthor>The PostgreSQL Global Development Group</corpauthor>
  &legal;
 </setinfo>

 &tutorial;
 &user;
 &admin;
 &programmer;
 &reference;
 &developer; 
 &setindex;

</set>

<!-- Keep this comment at the end of the file
Local variables:
mode:sgml
sgml-omittag:nil
sgml-shorttag:t
sgml-minimize-attributes:nil
sgml-always-quote-attributes:t
sgml-indent-step:1
sgml-indent-tabs-mode:nil
sgml-indent-data:t
sgml-parent-document:nil
sgml-default-dtd-file:"./reference.ced"
sgml-exposed-tags:nil
sgml-local-catalogs:("/usr/share/sgml/catalog")
sgml-local-ecat-files:nil
End:
-->
