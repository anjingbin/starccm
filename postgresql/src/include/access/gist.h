/*-------------------------------------------------------------------------
 *
 * gist.h
 *	  common declarations for the GiST access method code.
 *
 *
 * Portions Copyright (c) 1996-2001, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * $Id: gist.h,v 1.2 2004/07/05 08:55:54 SuLiang Exp $
 *
 *-------------------------------------------------------------------------
 */
#ifndef GIST_H
#define GIST_H

#include "access/itup.h"
#include "access/relscan.h"
#include "access/sdir.h"
#include "access/xlog.h"

/*
 * You can have as many strategies as you please in GiSTs,
 * as long as your consistent method can handle them.
 * The system doesn't really care what they are.
 */
#define GISTNStrategies					100

/*
 * amproc indexes for GiST indexes.
 */
#define GIST_CONSISTENT_PROC			1
#define GIST_UNION_PROC					2
#define GIST_COMPRESS_PROC				3
#define GIST_DECOMPRESS_PROC			4
#define GIST_PENALTY_PROC				5
#define GIST_PICKSPLIT_PROC				6
#define GIST_EQUAL_PROC					7
#define GISTNProcs						7


/*
 * Page opaque data in a GiST index page.
 */
#define F_LEAF			(1 << 0)

typedef struct GISTPageOpaqueData
{
	uint32		flags;
} GISTPageOpaqueData;

typedef GISTPageOpaqueData *GISTPageOpaque;

#define GIST_LEAF(entry) (((GISTPageOpaque) PageGetSpecialPointer((entry)->page))->flags & F_LEAF)

/*
 *	When we descend a tree, we keep a stack of parent pointers.
 */
typedef struct GISTSTACK
{
	struct GISTSTACK *gs_parent;
	OffsetNumber gs_child;
	BlockNumber gs_blk;
} GISTSTACK;

typedef struct GISTSTATE
{
	FmgrInfo	consistentFn[INDEX_MAX_KEYS];
	FmgrInfo	unionFn[INDEX_MAX_KEYS];
	FmgrInfo	compressFn[INDEX_MAX_KEYS];
	FmgrInfo	decompressFn[INDEX_MAX_KEYS];
	FmgrInfo	penaltyFn[INDEX_MAX_KEYS];
	FmgrInfo	picksplitFn[INDEX_MAX_KEYS];
	FmgrInfo	equalFn[INDEX_MAX_KEYS];

	TupleDesc	tupdesc;
} GISTSTATE;

#define isAttByVal( gs, anum ) (gs)->tupdesc->attrs[anum]->attbyval

/*
 *	When we're doing a scan, we need to keep track of the parent stack
 *	for the marked and current items.
 */
typedef struct GISTScanOpaqueData
{
	struct GISTSTACK *s_stack;
	struct GISTSTACK *s_markstk;
	uint16		s_flags;
	struct GISTSTATE *giststate;
} GISTScanOpaqueData;

typedef GISTScanOpaqueData *GISTScanOpaque;

/*
 *	When we're doing a scan and updating a tree at the same time, the
 *	updates may affect the scan.  We use the flags entry of the scan's
 *	opaque space to record our actual position in response to updates
 *	that we can't handle simply by adjusting pointers.
 */
#define GS_CURBEFORE	((uint16) (1 << 0))
#define GS_MRKBEFORE	((uint16) (1 << 1))

/* root page of a gist */
#define GISTP_ROOT				0

/*
 *	When we update a relation on which we're doing a scan, we need to
 *	check the scan and fix it if the update affected any of the pages it
 *	touches.  Otherwise, we can miss records that we should see.  The only
 *	times we need to do this are for deletions and splits.	See the code in
 *	gistscan.c for how the scan is fixed. These two constants tell us what sort
 *	of operation changed the index.
 */
#define GISTOP_DEL		0
#define GISTOP_SPLIT	1

/*
 * This is the Split Vector to be returned by the PickSplit method.
 */
typedef struct GIST_SPLITVEC
{
	OffsetNumber *spl_left;		/* array of entries that go left */
	int			spl_nleft;		/* size of this array */
	Datum		spl_ldatum;		/* Union of keys in spl_left */
	Datum		spl_lattr[INDEX_MAX_KEYS];		/* Union of subkeys in
												 * spl_left */
	int			spl_lattrsize[INDEX_MAX_KEYS];
	bool		spl_lisnull[INDEX_MAX_KEYS];

	OffsetNumber *spl_right;	/* array of entries that go right */
	int			spl_nright;		/* size of the array */
	Datum		spl_rdatum;		/* Union of keys in spl_right */
	Datum		spl_rattr[INDEX_MAX_KEYS];		/* Union of subkeys in
												 * spl_right */
	int			spl_rattrsize[INDEX_MAX_KEYS];
	bool		spl_risnull[INDEX_MAX_KEYS];

	int		   *spl_idgrp;
	int		   *spl_ngrp;		/* number in each group */
	char	   *spl_grpflag;	/* flags of each group */
} GIST_SPLITVEC;

/*
 * An entry on a GiST node.  Contains the key, as well as
 * its own location (rel,page,offset) which can supply the matching
 * pointer.  The size of the key is in bytes, and leafkey is a flag to
 * tell us if the entry is in a leaf node.
 */
typedef struct GISTENTRY
{
	Datum		key;
	Relation	rel;
	Page		page;
	OffsetNumber offset;
	int			bytes;
	bool		leafkey;
} GISTENTRY;

/*
 * macro to initialize a GISTENTRY
 */
#define gistentryinit(e, k, r, pg, o, b, l) \
	do { (e).key = (k); (e).rel = (r); (e).page = (pg); \
		 (e).offset = (o); (e).bytes = (b); (e).leafkey = (l); } while (0)

/* gist.c */
extern Datum gistbuild(PG_FUNCTION_ARGS);
extern Datum gistinsert(PG_FUNCTION_ARGS);
extern Datum gistbulkdelete(PG_FUNCTION_ARGS);
extern void _gistdump(Relation r);
extern void gistfreestack(GISTSTACK *s);
extern void initGISTstate(GISTSTATE *giststate, Relation index);
extern void freeGISTstate(GISTSTATE *giststate);
extern void gistdentryinit(GISTSTATE *giststate, int nkey, GISTENTRY *e,
			   Datum k, Relation r, Page pg, OffsetNumber o,
			   int b, bool l, bool isNull);
extern StrategyNumber RelationGetGISTStrategy(Relation, AttrNumber,
						RegProcedure);

extern void gist_redo(XLogRecPtr lsn, XLogRecord *record);
extern void gist_undo(XLogRecPtr lsn, XLogRecord *record);
extern void gist_desc(char *buf, uint8 xl_info, char *rec);

/* gistget.c */
extern Datum gistgettuple(PG_FUNCTION_ARGS);

#endif   /* GIST_H */
