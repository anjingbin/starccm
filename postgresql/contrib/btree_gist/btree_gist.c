#include "postgres.h"

#include "access/gist.h"
#include "access/itup.h"
#include "access/nbtree.h"

#include "utils/palloc.h"
#include "utils/geo_decls.h"
#include "utils/elog.h"

typedef int (*CMPFUNC) (const void *a, const void *b);
typedef void (*BINARY_UNION) (Datum *, char *);

typedef struct intkey
{
	int4		lower;
	int4		upper;
}	INT4KEY;

typedef struct tskey
{
	Timestamp	lower;
	Timestamp	upper;
}	TSKEY;

/* used for sorting */
typedef struct rix
{
	int			index;
	char	   *r;
}	RIX;

/*
** int4key in/out
*/
PG_FUNCTION_INFO_V1(int4key_in);
PG_FUNCTION_INFO_V1(int4key_out);
Datum		int4key_in(PG_FUNCTION_ARGS);
Datum		int4key_out(PG_FUNCTION_ARGS);

/*
** tskey in/out
*/
PG_FUNCTION_INFO_V1(tskey_in);
PG_FUNCTION_INFO_V1(tskey_out);
Datum		tskey_in(PG_FUNCTION_ARGS);
Datum		tskey_out(PG_FUNCTION_ARGS);

/*
** int4 ops
*/
PG_FUNCTION_INFO_V1(gint4_compress);
PG_FUNCTION_INFO_V1(gint4_union);
PG_FUNCTION_INFO_V1(gint4_picksplit);
PG_FUNCTION_INFO_V1(gint4_consistent);
PG_FUNCTION_INFO_V1(gint4_penalty);
PG_FUNCTION_INFO_V1(gint4_same);

Datum		gint4_compress(PG_FUNCTION_ARGS);
Datum		gint4_union(PG_FUNCTION_ARGS);
Datum		gint4_picksplit(PG_FUNCTION_ARGS);
Datum		gint4_consistent(PG_FUNCTION_ARGS);
Datum		gint4_penalty(PG_FUNCTION_ARGS);
Datum		gint4_same(PG_FUNCTION_ARGS);

static void gint4_binary_union(Datum *r1, char *r2);
static int	int4key_cmp(const void *a, const void *b);

/*
** timestamp ops
*/
PG_FUNCTION_INFO_V1(gts_compress);
PG_FUNCTION_INFO_V1(gts_union);
PG_FUNCTION_INFO_V1(gts_picksplit);
PG_FUNCTION_INFO_V1(gts_consistent);
PG_FUNCTION_INFO_V1(gts_penalty);
PG_FUNCTION_INFO_V1(gts_same);

Datum		gts_compress(PG_FUNCTION_ARGS);
Datum		gts_union(PG_FUNCTION_ARGS);
Datum		gts_picksplit(PG_FUNCTION_ARGS);
Datum		gts_consistent(PG_FUNCTION_ARGS);
Datum		gts_penalty(PG_FUNCTION_ARGS);
Datum		gts_same(PG_FUNCTION_ARGS);

static void gts_binary_union(Datum *r1, char *r2);
static int	tskey_cmp(const void *a, const void *b);

#define TimestampGetDatumFast(X) Float8GetDatumFast(X)

/* define for comparison */
#define TSGE( ts1, ts2 ) (DatumGetBool(DirectFunctionCall2( \
		timestamp_ge, \
		PointerGetDatum( ts1 ), \
		PointerGetDatum( ts2 ) \
)))
#define TSGT( ts1, ts2 ) (DatumGetBool(DirectFunctionCall2( \
		timestamp_gt, \
		PointerGetDatum( ts1 ), \
		PointerGetDatum( ts2 ) \
)))
#define TSEQ( ts1, ts2 ) (DatumGetBool(DirectFunctionCall2( \
		timestamp_eq, \
		PointerGetDatum( ts1 ), \
		PointerGetDatum( ts2 ) \
)))
#define TSLT( ts1, ts2 ) (DatumGetBool(DirectFunctionCall2( \
		timestamp_lt, \
		PointerGetDatum( ts1 ), \
		PointerGetDatum( ts2 ) \
)))
#define TSLE( ts1, ts2 ) (DatumGetBool(DirectFunctionCall2( \
		timestamp_le, \
		PointerGetDatum( ts1 ), \
		PointerGetDatum( ts2 ) \
)))

/*
** Common btree-function (for all ops)
*/
static GIST_SPLITVEC *btree_picksplit(bytea *entryvec, GIST_SPLITVEC *v,
				BINARY_UNION bu, CMPFUNC cmp);

PG_FUNCTION_INFO_V1(btree_decompress);
Datum		btree_decompress(PG_FUNCTION_ARGS);

/**************************************************
 * int4 ops
 **************************************************/

Datum
gint4_compress(PG_FUNCTION_ARGS)
{
	GISTENTRY  *entry = (GISTENTRY *) PG_GETARG_POINTER(0);
	GISTENTRY  *retval;

	if (entry->leafkey)
	{
		INT4KEY    *r = palloc(sizeof(INT4KEY));

		retval = palloc(sizeof(GISTENTRY));
		r->lower = r->upper = (entry->key);

		gistentryinit(*retval, PointerGetDatum(r), entry->rel, entry->page,
					  entry->offset, sizeof(INT4KEY), FALSE);

	}
	else
		retval = entry;
	PG_RETURN_POINTER(retval);
}

Datum
gint4_consistent(PG_FUNCTION_ARGS)
{
	GISTENTRY  *entry = (GISTENTRY *) PG_GETARG_POINTER(0);
	int4		query = PG_GETARG_INT32(1);
	INT4KEY    *kkk = (INT4KEY *) DatumGetPointer(entry->key);
	StrategyNumber strategy = (StrategyNumber) PG_GETARG_UINT16(2);
	bool		retval;

	switch (strategy)
	{
		case BTLessEqualStrategyNumber:
			retval = (query >= kkk->lower);
			break;
		case BTLessStrategyNumber:
			if (GIST_LEAF(entry))
				retval = (query > kkk->lower);
			else
				retval = (query >= kkk->lower);
			break;
		case BTEqualStrategyNumber:
			/* in leaf page kkk->lower always = kkk->upper */
			if (GIST_LEAF(entry))
				retval = (query == kkk->lower);
			else
				retval = (kkk->lower <= query && query <= kkk->upper);
			break;
		case BTGreaterStrategyNumber:
			if (GIST_LEAF(entry))
				retval = (query < kkk->upper);
			else
				retval = (query <= kkk->upper);
			break;
		case BTGreaterEqualStrategyNumber:
			retval = (query <= kkk->upper);
			break;
		default:
			retval = FALSE;
	}
	PG_RETURN_BOOL(retval);
}

Datum
gint4_union(PG_FUNCTION_ARGS)
{
	bytea	   *entryvec = (bytea *) PG_GETARG_POINTER(0);
	int			i,
				numranges;
	INT4KEY    *cur,
			   *out = palloc(sizeof(INT4KEY));

	numranges = (VARSIZE(entryvec) - VARHDRSZ) / sizeof(GISTENTRY);
	*(int *) PG_GETARG_POINTER(1) = sizeof(INT4KEY);

	cur = (INT4KEY *) DatumGetPointer((((GISTENTRY *) (VARDATA(entryvec)))[0].key));
	out->lower = cur->lower;
	out->upper = cur->upper;

	for (i = 1; i < numranges; i++)
	{
		cur = (INT4KEY *) DatumGetPointer((((GISTENTRY *) (VARDATA(entryvec)))[i].key));
		if (out->lower > cur->lower)
			out->lower = cur->lower;
		if (out->upper < cur->upper)
			out->upper = cur->upper;
	}

	PG_RETURN_POINTER(out);
}

Datum
gint4_penalty(PG_FUNCTION_ARGS)
{
	INT4KEY    *origentry = (INT4KEY *) DatumGetPointer(((GISTENTRY *) PG_GETARG_POINTER(0))->key);
	INT4KEY    *newentry = (INT4KEY *) DatumGetPointer(((GISTENTRY *) PG_GETARG_POINTER(1))->key);
	float	   *result = (float *) PG_GETARG_POINTER(2);

	*result = Max(newentry->upper - origentry->upper, 0) +
		Max(origentry->lower - newentry->lower, 0);

	PG_RETURN_POINTER(result);
}

Datum
gint4_picksplit(PG_FUNCTION_ARGS)
{
	PG_RETURN_POINTER(btree_picksplit(
									  (bytea *) PG_GETARG_POINTER(0),
								  (GIST_SPLITVEC *) PG_GETARG_POINTER(1),
									  gint4_binary_union,
									  int4key_cmp
									  ));
}

Datum
gint4_same(PG_FUNCTION_ARGS)
{
	INT4KEY    *b1 = (INT4KEY *) PG_GETARG_POINTER(0);
	INT4KEY    *b2 = (INT4KEY *) PG_GETARG_POINTER(1);
	bool	   *result = (bool *) PG_GETARG_POINTER(2);

	*result = (b1->lower == b2->lower && b1->upper == b2->upper) ? TRUE : FALSE;
	PG_RETURN_POINTER(result);
}

static void
gint4_binary_union(Datum *r1, char *r2)
{
	INT4KEY    *b1;
	INT4KEY    *b2 = (INT4KEY *) r2;

	if (!DatumGetPointer(*r1))
	{
		*r1 = PointerGetDatum(palloc(sizeof(INT4KEY)));
		b1 = (INT4KEY *) DatumGetPointer(*r1);
		b1->upper = b2->upper;
		b1->lower = b2->lower;
	}
	else
	{
		b1 = (INT4KEY *) DatumGetPointer(*r1);

		b1->lower = (b1->lower > b2->lower) ?
			b2->lower : b1->lower;
		b1->upper = (b1->upper > b2->upper) ?
			b1->upper : b2->upper;
	}
}


static int
int4key_cmp(const void *a, const void *b)
{
	return (((INT4KEY *) (((RIX *) a)->r))->lower - ((INT4KEY *) (((RIX *) b)->r))->lower);
}

/**************************************************
 * timestamp ops
 **************************************************/

Datum
gts_compress(PG_FUNCTION_ARGS)
{
	GISTENTRY  *entry = (GISTENTRY *) PG_GETARG_POINTER(0);
	GISTENTRY  *retval;

	if (entry->leafkey)
	{
		TSKEY	   *r = (TSKEY *) palloc(sizeof(TSKEY));
		retval = palloc(sizeof(GISTENTRY));
		r->lower = r->upper = *(Timestamp *) (entry->key);
		gistentryinit(*retval, PointerGetDatum(r),
			  entry->rel, entry->page,
			  entry->offset, sizeof(TSKEY), FALSE);
	}
	else
		retval = entry;
	PG_RETURN_POINTER(retval);
}

Datum
gts_consistent(PG_FUNCTION_ARGS)
{
	GISTENTRY  *entry = (GISTENTRY *) PG_GETARG_POINTER(0);
	Timestamp  *query = (Timestamp *) PG_GETARG_POINTER(1);
	StrategyNumber strategy = (StrategyNumber) PG_GETARG_UINT16(2);
	bool		retval;
	TSKEY	   *key;

	/*
	 * * if entry is not leaf, use gbox_internal_consistent, * else use
	 * gbox_leaf_consistent
	 */
	if (!entry->key)
		return FALSE;
	key = (TSKEY *) DatumGetPointer(entry->key);

	switch (strategy)
	{
		case BTLessEqualStrategyNumber:
			retval = TSGE(query, &(key->lower));
			break;
		case BTLessStrategyNumber:
			if (GIST_LEAF(entry))
				retval = TSGT(query, &(key->lower));
			else
				retval = TSGE(query, &(key->lower));
			break;
		case BTEqualStrategyNumber:
			/* in leaf page key->lower always = key->upper */
			if (GIST_LEAF(entry))
				retval = TSEQ(query, &(key->lower));
			else
				retval = (TSLE(&(key->lower), query) && TSLE(query, &(key->upper)));
			break;
		case BTGreaterStrategyNumber:
			if (GIST_LEAF(entry))
				retval = TSLT(query, &(key->upper));
			else
				retval = TSLE(query, &(key->upper));
			break;
		case BTGreaterEqualStrategyNumber:
			retval = TSLE(query, &(key->upper));
			break;
		default:
			retval = FALSE;
	}
	PG_RETURN_BOOL(retval);
}

Datum
gts_union(PG_FUNCTION_ARGS)
{
	bytea	   *entryvec = (bytea *) PG_GETARG_POINTER(0);
	int			i,
				numranges;
	TSKEY	   *cur,
			   *out = palloc(sizeof(TSKEY));

	numranges = (VARSIZE(entryvec) - VARHDRSZ) / sizeof(GISTENTRY);
	*(int *) PG_GETARG_POINTER(1) = sizeof(TSKEY);

	cur = (TSKEY *) DatumGetPointer((((GISTENTRY *) (VARDATA(entryvec)))[0].key));
	out->lower = cur->lower;
	out->upper = cur->upper;

	for (i = 1; i < numranges; i++)
	{
		cur = (TSKEY *) DatumGetPointer((((GISTENTRY *) (VARDATA(entryvec)))[i].key));
		if (TSGT(&out->lower, &cur->lower))
			out->lower = cur->lower;
		if (TSLT(&out->upper, &cur->upper))
			out->upper = cur->upper;
	}

	PG_RETURN_POINTER(out);
}

Datum
gts_penalty(PG_FUNCTION_ARGS)
{
	TSKEY	   *origentry = (TSKEY *) DatumGetPointer(((GISTENTRY *) PG_GETARG_POINTER(0))->key);
	TSKEY	   *newentry = (TSKEY *) DatumGetPointer(((GISTENTRY *) PG_GETARG_POINTER(1))->key);
	float	   *result = (float *) PG_GETARG_POINTER(2);
	Interval   *intr;

	intr = DatumGetIntervalP(DirectFunctionCall2(
												 timestamp_mi,
									  TimestampGetDatumFast(newentry->upper),
								   TimestampGetDatumFast(origentry->upper)));

	/* see interval_larger */
	*result = Max(intr->time + intr->month * (30.0 * 86400), 0);
	pfree(intr);

	intr = DatumGetIntervalP(DirectFunctionCall2(
												 timestamp_mi,
									 TimestampGetDatumFast(origentry->lower),
									TimestampGetDatumFast(newentry->lower)));

	/* see interval_larger */
	*result += Max(intr->time + intr->month * (30.0 * 86400), 0);
	pfree(intr);

	PG_RETURN_POINTER(result);
}

Datum
gts_picksplit(PG_FUNCTION_ARGS)
{
	PG_RETURN_POINTER(btree_picksplit(
									  (bytea *) PG_GETARG_POINTER(0),
								  (GIST_SPLITVEC *) PG_GETARG_POINTER(1),
									  gts_binary_union,
									  tskey_cmp
									  ));
}

Datum
gts_same(PG_FUNCTION_ARGS)
{
	TSKEY	   *b1 = (TSKEY *) PG_GETARG_POINTER(0);
	TSKEY	   *b2 = (TSKEY *) PG_GETARG_POINTER(1);

	bool	   *result = (bool *) PG_GETARG_POINTER(2);

	if (b1 && b2)
		*result = (TSEQ(&(b1->lower), &(b2->lower)) && TSEQ(&(b1->upper), &(b2->upper))) ? TRUE : FALSE;
	else
		*result = (b1 == NULL && b2 == NULL) ? TRUE : FALSE;
	PG_RETURN_POINTER(result);
}

static void
gts_binary_union(Datum *r1, char *r2)
{
	TSKEY	   *b1;
	TSKEY	   *b2 = (TSKEY *) r2;

	if (!DatumGetPointer(*r1))
	{
		*r1 = PointerGetDatum(palloc(sizeof(TSKEY)));
		b1 = (TSKEY *) DatumGetPointer(*r1);
		b1->upper = b2->upper;
		b1->lower = b2->lower;
	}
	else
	{
		b1 = (TSKEY *) DatumGetPointer(*r1);

		b1->lower = (TSGT(&b1->lower, &b2->lower)) ?
			b2->lower : b1->lower;
		b1->upper = (TSGT(&b1->upper, &b2->upper)) ?
			b1->upper : b2->upper;
	}
}

static int
tskey_cmp(const void *a, const void *b)
{
	return DatumGetInt32(
						 DirectFunctionCall2(
											 timestamp_cmp,
				  TimestampGetDatumFast(((TSKEY *) (((RIX *) a)->r))->lower),
				   TimestampGetDatumFast(((TSKEY *) (((RIX *) b)->r))->lower)
											 )
		);
}

/**************************************************
 * Common btree-function (for all ops)
 **************************************************/

/*
** The GiST PickSplit method
*/
static GIST_SPLITVEC *
btree_picksplit(bytea *entryvec, GIST_SPLITVEC *v, BINARY_UNION bu, CMPFUNC cmp)
{
	OffsetNumber i;
	RIX		   *array;
	OffsetNumber maxoff;
	int			nbytes;

	maxoff = ((VARSIZE(entryvec) - VARHDRSZ) / sizeof(GISTENTRY)) - 1;
	nbytes = (maxoff + 2) * sizeof(OffsetNumber);
	v->spl_left = (OffsetNumber *) palloc(nbytes);
	v->spl_right = (OffsetNumber *) palloc(nbytes);
	v->spl_nleft = 0;
	v->spl_nright = 0;
	v->spl_ldatum = PointerGetDatum(0);
	v->spl_rdatum = PointerGetDatum(0);
	array = (RIX *) palloc(sizeof(RIX) * (maxoff + 1));

	/* copy the data into RIXes, and sort the RIXes */
	for (i = FirstOffsetNumber; i <= maxoff; i = OffsetNumberNext(i))
	{
		array[i].index = i;
		array[i].r = (char *) DatumGetPointer((((GISTENTRY *) (VARDATA(entryvec)))[i].key));
	}
	qsort((void *) &array[FirstOffsetNumber], maxoff - FirstOffsetNumber + 1,
		  sizeof(RIX), cmp);

	for (i = FirstOffsetNumber; i <= maxoff; i = OffsetNumberNext(i))
	{
		if (i <= (maxoff - FirstOffsetNumber + 1) / 2)
		{
			v->spl_left[v->spl_nleft] = array[i].index;
			v->spl_nleft++;
			(*bu) (&v->spl_ldatum, array[i].r);
		}
		else
		{
			v->spl_right[v->spl_nright] = array[i].index;
			v->spl_nright++;
			(*bu) (&v->spl_rdatum, array[i].r);
		}
	}
	pfree(array);

	return (v);
}

/*
** GiST DeCompress methods
** do not do anything.
*/
Datum
btree_decompress(PG_FUNCTION_ARGS)
{
	PG_RETURN_POINTER(PG_GETARG_POINTER(0));
}


/**************************************************
 * In/Out for keys, not really needed
 **************************************************/
Datum
int4key_in(PG_FUNCTION_ARGS)
{
	INT4KEY    *key = palloc(sizeof(INT4KEY));

	if (sscanf(PG_GETARG_POINTER(0), "%d|%d", &(key->lower), &(key->upper)) != 2)
		elog(ERROR, "Error in input format");

	PG_RETURN_POINTER(key);
}

Datum
int4key_out(PG_FUNCTION_ARGS)
{
	INT4KEY    *key = (INT4KEY *) PG_GETARG_POINTER(0);
	char	   *str = palloc(sizeof(char) * 22);

	sprintf(str, "%d|%d", key->lower, key->upper);
	PG_RETURN_POINTER(str);
}

Datum
tskey_in(PG_FUNCTION_ARGS)
{
	elog(ERROR, "Not implemented");
	PG_RETURN_POINTER(NULL);
}

Datum
tskey_out(PG_FUNCTION_ARGS)
{
	elog(ERROR, "Not implemented");
	PG_RETURN_POINTER(NULL);
}
