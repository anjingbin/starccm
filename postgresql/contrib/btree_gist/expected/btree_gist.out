--
-- first, define the datatype.  Turn off echoing so that expected file
-- does not depend on contents of seg.sql.
--
\set ECHO none
create table inttmp (b int4);
\copy inttmp from 'data/test_btree.data'
create table tstmp ( t timestamp without time zone );
\copy tstmp from 'data/test_btree_ts.data'
-- without idx
select count(*) from inttmp where b <=10;
 count 
-------
    11
(1 row)

select count(*) from tstmp where t < '2001-05-29 08:33:09';
 count 
-------
    66
(1 row)

-- create idx
create index aaaidx on inttmp using gist ( b );
create index tsidx on tstmp using gist ( t );
--with idx
set enable_seqscan=off;
select count(*) from inttmp where b <=10;
 count 
-------
    11
(1 row)

select count(*) from tstmp where t < '2001-05-29 08:33:09';
 count 
-------
    66
(1 row)

